src/NITTA/Intermediate/ProcessWaves.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- |
+Module      : NITTA.Intermediate.ProcessWaves
+Description : Process waves graph
+Copyright   : (c) Aleksandr Penskoi, 2022
+License     : BSD3
+Maintainer  : aleksandr.penskoi@gmail.com
+Stability   : experimental
+-}
+module NITTA.Intermediate.ProcessWaves (
+    ProcessWave (..),
+    buildProcessWaves,
+    reorderAlgorithm,
+    estimateWaves,
+) where
+
+import Data.Data
+import qualified Data.List as L
+import qualified Data.List.Split as L
+import qualified Data.Map as M
+import qualified Data.Set as S
+import Data.String.ToString
+import GHC.Generics
+import NITTA.Intermediate.Functions
+import NITTA.Intermediate.Types
+import NITTA.Utils.Base
+
+data ProcessWave v x = ProcessWave
+    { -- |Functions that can be executed at this wave
+      pwFs :: [F v x]
+    , -- |Set of output variables related to the functions from this step
+      pwOut :: S.Set String
+    }
+    deriving (Show, Generic)
+
+data Builder v x = Builder
+    { -- |Functions that can be calculated due to lack of ready input values
+      pwRemains :: S.Set (F v x)
+    , -- |Variables that defined at the beginning of the process
+      pwIn :: S.Set String
+    , -- |Variables that is ready to be used us inputs
+      pwReadyIn :: S.Set String
+    , -- |Resulting process flow
+      pwGraph :: [ProcessWave v x]
+    }
+    deriving (Show, Generic)
+
+{- |
+Sort functions in order of execution.
+Note that some function could be executed in parallel, in this case we save order from the original list.
+-}
+reorderAlgorithm :: (Ord v, ToString v, Typeable x, Typeable v) => [F v x] -> [F v x]
+reorderAlgorithm alg = concatMap (\ProcessWave{pwFs} -> pwFs) $ buildProcessWaves [] alg
+
+{- |
+Functions can be divided into waves of execution.
+For each output variable, we define the wave number on which the variable will be defined.
+-}
+estimateWaves :: (Ord v, Num t, ToString v, Typeable x, Typeable v) => [F v x] -> [v] -> M.Map v t
+estimateWaves fs alreadyVars = M.fromList $ go 0 $ buildProcessWaves alreadyVars fs
+    where
+        go n (ProcessWave{pwFs} : pss) = go (n + 1) pss <> [(out, n) | out <- S.toList $ unionsMap outputs pwFs]
+        go _ [] = []
```

Comment:
Do you mean the name of the recursion function? I thought that go is common name for haskell because it's used  [GHC.Base module](https://hackage.haskell.org/package/base-4.15.0.0/docs/src/GHC-Base.html#%2B%2B)